Add http-lb resolver

mberhault opened the pull request:
To be used with http-only load balancers.
We get http(s)://<load-balancer>/_status_nodes and extract the first node address to use as the gossip address.

There's some oddness I want to refactor before getting this in. Mainly, in single-node (or first node) setups, we will be talking to ourselves. We need to make sure the http server is up while this resolver fails. This is where the separate http client and exhausted logic comes in.

Moved resolver to base/ to avoid base<->util import cycle.

Commit:
Add http-lb resolver

tschottdorf commented on base/resolver_node_lookup.go:
```diff
@@ -0,0 +1,127 @@
+
+// nodeLookupResolver implements Resolver.
+// It queries http(s)://<address>/_status/nodes and extracts the first node's address.
+// This is useful for http load balancers which will not foward RPC.
```
s/fow/for/

MBerhault replied:
ugh! I always do this. Fixed.

tschottdorf commented on base/resolver_node_lookup.go:
```diff
@@ -0,0 +1,127 @@
+func (nl *nodeLookupResolver) GetAddress() (net.Addr, error) {
+	if nl.exhausted {
+		nl.exhausted = false
+		return nil, util.Errorf("skipping temporarily-exhausted resolved")
```
s/Errorf/Error/.
Also, I don't get the message.

tschottdorf commented on base/resolver_node_lookup.go:
```diff
@@ -0,0 +1,127 @@
+func (nl *nodeLookupResolver) Addr() string { return nl.addr }
+
+// GetAddress returns a net.Addr or error.
+// Upon errors, we set exhausted=true, then flip it back when called again.
```
why this dance with setting exhausted? What's the upshot?

MBerhault replied:
It's because of the weird startup issue in single-node mode. The http server doesn't start until we've either resolved a gossip bootstrap address, or exhausted them all. I get around that by setting the resolver to exhausted on errors, and returning right away if we are exhausted. The caller runs GetAddress() before checking IsExhausted().

I still want to rework this part, it's getting really messy.

tschottdorf replied:
Ok.

bdarnell replied:
Put this explanation (that we have to set the exhausted flag to allow the HTTP server to start up) in a comment in the code.

MBerhault replied:
Done.

tschottdorf commented on server/cli/flags.go:
```diff
@@ -71,9 +71,13 @@ var flagUsage = map[string]string{
+          or "self" for single-node systems.
+        - unix: unix socket
+        - lb: RPC load balancer fowarding to an arbitrary node
+        - http-lb: HTTP load balancer: we query http(s)://lb/_status/nodes
```
how would you specify whether to use https? `http-lb=https://....?`

MBerhault replied:
you don't, we use the node security settings. The context (initialized from flags) is the one that decides whether we're using http or https and loads the TLS config.

Since the entire cluster needs to be secure or insecure, we can do this automatically.

tschottdorf replied:
ah, right. Makes  sense.

tschottdorf commented:
I just merged my #1010 fix so if you rebase hopefully it'll be greener.

BramGruneir commented:
Perhaps you should put the resolver back into gossip?  I moved it to util because it seemed liked it should exist next to UnresolvedAddr.  Now it's getting even more tied into gossip, perhaps that's where it should be.

spencerkimball commented:
I really dislike moving resolver to base. I'm missing why that's necessary...

spencerkimball commented on base/resolver_node_lookup.go:
```diff
@@ -0,0 +1,127 @@
+// Copyright 2015 The Cockroach Authors.
```
The naming on these files is odd. I think it would be more "standard" to name this "node_lookup_resolver.go" and the other "socket_resolver.go". Those would comport with the actual struct names.

MBerhault replied:
Yup. Ben proposed a gossip/resolver/ package which sounds much more sensible.
I'll move it there and rename the files to just plain "node_lookup" and "socket".

spencerkimball commented on base/resolver_node_lookup.go:
```diff
@@ -0,0 +1,127 @@
+
+	nl.exhausted = true
+	// TODO(marc): put common URIs in base and reuse everywhere.
+	url := fmt.Sprintf("%s://%s/_status/nodes", nl.context.RequestScheme(), nl.addr)
```
This doesn't feel like the correct endpoint for this. Correct me if I'm wrong, but you're running into an issue where you need to get the address of a node in the cluster via HTTP because the load balancer play nicely with HTTP forwarding. You then use the address to directly connect to a gossip node.

What you actually want to do is improve the `/_status/local` endpoint by adding the address of the node to the json which is currently returned (it currently just has the build info). You could then just make that HTTP request repeatedly and get a random sampling of the external IP addresses of the nodes in the cluster.

I'll send you a PR which provides the address there.

Going to `/_status/nodes` is problematic because most nodes won't be able to write updated information there without first using gossip to connect to the cluster, so they can get first range metadata, etc.

MBerhault replied:
Yup. I put it there for now since that's all there is to get node addresses.
I can add the /local endpoint instead, that would definitely be preferable, since right now I had to do some weird logic for single-node startup.

bdarnell commented:
We should probably have a `resolver` package (maybe put it under the gossip directory?) since this is growing into multiple files.

Commit:
Fix resolver package

Commit:
Query /_status/local to get node address

MBerhault commented:
Addressed review comments:
- move resolver files to gossip/resolver and rename
- query /_status/local for node address rather than /_status/nodes

I still have a single-node (or first node) startup issue since we a valid gossip address to start the http server.

MBerhault commented:
Oh, and the way I handle single-node or first-node startup is by passing:
--gossip=localhost:8080,http-lb=localhost:8080

My plan is to have the status pages start serving as soon as possible (and definitely before gossip is up).
This way, single node bootstrap by using http-lb will work just fine.
It means that whatever the load balancer uses for the health-check needs to return 200.

MBerhault commented:
This is done. Please take another look.

bdarnell commented:
LGTM

Commit:
Clarify exhaust logic
